lua/fzfx/cfg/_lsp_locations.lua

```diff
@@ -47,6 +47,9 @@ end
 --- @param renderer fun(text:string):string
 --- @return string?
 M._colorize_lsp_range = function(line, range, renderer)
+  if not line then
```

Comment:
hi @stong1994 ,

Can you use `str.empty(line)` to detect whether a `line` is nil or empty string?

Reply:
I have identified the issue. It appears there is a problem when executing the `vim.lsp.buf_request_sync` function(the 'method' param is "callHierarchy/outgoingCalls"), as the response I received is incorrect.

I have generated some logs to help diagnose the problem, and here is the output:
<img width="1576" alt="lsp1" src="https://github.com/linrongbin16/fzfx.nvim/assets/23580595/efdf2d85-a967-4385-aec7-df8d51bbed1e">


The crux of the issue is that the `sms_error.go` file contains only 24 lines of code. However, the fromRange in the response is pointing to lines 38 and 68, which is not possible given the file's length.

Reply:
thanks to your work!

can you give me a golang repository that can reproduce this issue? so I may also could help fix this issue.

Reply:
And, can you take a look at this line: https://github.com/linrongbin16/fzfx.nvim/blob/9b70ddb4026d93ad0a2fb9ad8b15ddfd6d4eae42/lua/fzfx/cfg/_lsp_locations.lua?plain=1#L398

It's the request body that sent to the lsp server, can you help verify if the request is correct?

Reply:
there are two calls to `lsp.buf_request_sync`, the first one has no problem, and the second one which logs at https://github.com/linrongbin16/fzfx.nvim/blob/9b70ddb4026d93ad0a2fb9ad8b15ddfd6d4eae42/lua/fzfx/cfg/_lsp_locations.lua?plain=1#L438 has a incorrect response, and I'll show you the log:
(I have changed the log code to ` log.err(
      "|_make_lsp_call_hierarchy_provider| 2nd call, opts:%s, lsp_item: %s, lsp_results2:%s, lsp_err2:%s, method: %s",
      vim.inspect(opts),
      vim.inspect(lsp_item),
      vim.inspect(lsp_results2),
      vim.inspect(lsp_err2),
      vim.inspect(opts.method)
    )`)
the output is:

> 2024-04-25 10:56:22,143649 [@/Users/stong/.local/share/nvim/lazy/fzfx.nvim/lua/fzfx/cfg/_lsp_locations.lua:438] ERROR: |_make_lsp_call_hierarchy_provider| 2nd call, opts:{
  capability = "callHierarchyProvider",
  method = "callHierarchy/outgoingCalls"
}, lsp_item: { {
    detail = "dm.sp.sms/utils/sms_utils • sender.go",
    kind = 12,
    name = "ytxSend",
    range = {
      ["end"] = {
        character = 12,
        line = 35
      },
      start = {
        character = 5,
        line = 35
      }
    },
    selectionRange = {
      ["end"] = {
        character = 12,
        line = 35
      },
      start = {
        character = 5,
        line = 35
      }
    },
    uri = "file:///Users/stong/Project/Company/dm.sp.sms/utils/sms_utils/sender.go"
  } }, lsp_results2:{
  [2] = {
    result = { {
        fromRanges = { {
            ["end"] = {
              character = 14,
              line = 71
            },
            start = {
              character = 5,
              line = 71
            }
          } },
        to = {
          detail = "dm.sp.sms/utils/sms_utils • sms.go",
          kind = 12,
          name = "SetSmsSid",
          range = {
            ["end"] = {
              character = 28,
              line = 361
            },
            start = {
              character = 19,
              line = 361
            }
          },
          selectionRange = {
            ["end"] = {
              character = 28,
              line = 361
            },
            start = {
              character = 19,
              line = 361
            }
          },
          uri = "file:///Users/stong/Project/Company/dm.sp.sms/utils/sms_utils/sms.go"
        }
      }, {
        fromRanges = { {
            ["end"] = {
              character = 26,
              line = 36
            },
            start = {
              character = 14,
              line = 36
            }
          } },
        to = {
          detail = "dm.sp.sms/utils/sms_utils • yuntongxun.go",
          kind = 12,
          name = "NewYTXSender",
          range = {
            ["end"] = {
              character = 17,
              line = 53
            },
            start = {
              character = 5,
              line = 53
            }
          },
          selectionRange = {
            ["end"] = {
              character = 17,
              line = 53
            },
            start = {
              character = 5,
              line = 53
            }
          },
          uri = "file:///Users/stong/Project/Company/dm.sp.sms/utils/sms_utils/yuntongxun.go"
        }
      }, {
        fromRanges = { {
            ["end"] = {
              character = 18,
              line = 38
            },
            start = {
              character = 9,
              line = 38
            }
          }, {
            ["end"] = {
              character = 18,
              line = 63
            },
            start = {
              character = 9,
              line = 63
            }
          }, {
            ["end"] = {
              character = 18,
              line = 68
            },
            start = {
              character = 9,
              line = 68
            }
          } },
        to = {
          detail = "dm.sp.sms/utils/sms_utils • sms_errors.go",
          kind = 12,
          name = "NewSmsErr",
          range = {
            ["end"] = {
              character = 14,
              line = 12
            },
            start = {
              character = 5,
              line = 12
            }
          },
          selectionRange = {
            ["end"] = {
              character = 14,
              line = 12
            },
            start = {
              character = 5,
              line = 12
            }
          },
          uri = "file:///Users/stong/Project/Company/dm.sp.sms/utils/sms_utils/sms_errors.go"
        }
      }, {
        fromRanges = { {
            ["end"] = {
              character = 22,
              line = 50
            },
            start = {
              character = 14,
              line = 50
            }
          } },
        to = {
          detail = "strings • strings.go",
          kind = 12,
          name = "Contains",
          range = {
            ["end"] = {
              character = 13,
              line = 59
            },
            start = {
              character = 5,
              line = 59
            }
          },
          selectionRange = {
            ["end"] = {
              character = 13,
              line = 59
            },
            start = {
              character = 5,
              line = 59
            }
          },
          uri = "file:///Users/stong/sdk/go1.21.3/src/strings/strings.go"
        }
      }, {
        fromRanges = { {
            ["end"] = {
              character = 30,
              line = 51
            },
            start = {
              character = 26,
              line = 51
            }
          }, {
            ["end"] = {
              character = 25,
              line = 67
            },
            start = {
              character = 21,
              line = 67
            }
          } },
        to = {
          detail = "strconv • atoi.go",
          kind = 12,
          name = "Atoi",
          range = {
            ["end"] = {
              character = 9,
              line = 246
            },
            start = {
              character = 5,
              line = 246
            }
          },
          selectionRange = {
            ["end"] = {
              character = 9,
              line = 246
            },
            start = {
              character = 5,
              line = 246
            }
          },
          uri = "file:///Users/stong/sdk/go1.21.3/src/strconv/atoi.go"
        }
      }, {
        fromRanges = { {
            ["end"] = {
              character = 36,
              line = 61
            },
            start = {
              character = 23,
              line = 61
            }
          } },
        to = {
          detail = "dm.sp.sms/utils/sms_utils • yuntongxun.go",
          kind = 12,
          name = "TplSingleSend",
          range = {
            ["end"] = {
              character = 35,
              line = 98
            },
            start = {
              character = 22,
              line = 98
            }
          },
          selectionRange = {
            ["end"] = {
              character = 35,
              line = 98
            },
            start = {
              character = 22,
              line = 98
            }
          },
          uri = "file:///Users/stong/Project/Company/dm.sp.sms/utils/sms_utils/yuntongxun.go"
        }
      }, {
        fromRanges = { {
            ["end"] = {
              character = 22,
              line = 66
            },
            start = {
              character = 16,
              line = 66
            }
          } },
        to = {
          detail = "dm.sp.sms/utils/sms_utils • yuntongxun.go",
          kind = 12,
          name = "IsSucc",
          range = {
            ["end"] = {
              character = 29,
              line = 49
            },
            start = {
              character = 23,
              line = 49
            }
          },
          selectionRange = {
            ["end"] = {
              character = 29,
              line = 49
            },
            start = {
              character = 23,
              line = 49
            }
          },
          uri = "file:///Users/stong/Project/Company/dm.sp.sms/utils/sms_utils/yuntongxun.go"
        }
      } }
  }
}, lsp_err2:nil, method: "callHierarchy/outgoingCalls"

Reply:
The code above is part of my company's project, so I cannot share it. However, I am attempting to use a public repository to replicate the issue.
BTW, I suspect this may be a bug with the lsp server I am utilizing. Therefore, I am considering trying an alternative server

Reply:
Sadly, I'm not sure why this happens.

The `callHierarchy/outgoingCalls` needs two APIs:
1. use `"textDocument/prepareCallHierarchy"` to prepare the hierarchy info.
2. use `callHierarchy/outgoingCalls` to query outgoing calls.

Based on your logs, the error exists in the 2nd API results.

Reply:
> BTW, I suspect this may be a bug with the lsp server I am utilizing. Therefore, I am considering trying an alternative server

In recent days, I found the LSP protocol added a new protocol `didChangeWatchedFiles` (see: [reddit - Can someone explain the benefit of Neovim 0.10 LSP didChangeWatchedFiles](https://www.reddit.com/r/neovim/comments/1bhh7dj/can_someone_explain_the_benefit_of_neovim_010_lsp/)), I'm not sure if it's related to our issue.

Reply:
hi @stong1994 ,

after checked the logic of `_colorize_lsp_range` function, and there're two places calls this function:

1. https://github.com/linrongbin16/fzfx.nvim/blob/352906f5aba7f79503524ecd1c6efa5c28f97cc0/lua/fzfx/cfg/_lsp_locations.lua?plain=1#L131
2. https://github.com/linrongbin16/fzfx.nvim/blob/352906f5aba7f79503524ecd1c6efa5c28f97cc0/lua/fzfx/cfg/_lsp_locations.lua?plain=1#L322

They're to rendering the colors for these lines.

So I think a better way is to check if the `r` (the range) is existing in the `filelines`, this is actually a index check in lua array, correct?

If the `r` range is even not existed in the `filelines`, then we can simply skip it, instead of call the `_colorize_lsp_range` function and return a `nil` value. Because the returned `item_line` is still need to be rendered and print in the final results.

Reply:
This is not related to the issue, as I have not modified any files. I suspect the root cause may be related to the Go LSP, and I intend to investigate further. By the way, I am aware of the `str.empty` API and have chosen to use it in place of the expression `not str.not_empty`.

Reply:
hi @linrongbin16 
I noticed that Lua doesn't have a 'continue' statement for 'for' loops. Should I revise the code like this?
<img width="1052" alt="image" src="https://github.com/linrongbin16/fzfx.nvim/assets/23580595/cfa00b22-73e5-4fc4-952e-65baf2cd1155">

Reply:
> I am aware of the `str.empty` API.

That's better choice.

> I noticed that Lua doesn't have a 'continue' statement for 'for' loops. Should I revise the code like this?

It looks good

Reply:
hi @stong1994 ,

Once you had tested these changes, please select the related items in the "Regresion test" check list, to make sure they're working.

Reply:
hi @linrongbin16 
The 'Ctrl+J' and 'Ctrl+K' shortcuts are not working. Could I be missing some configuration settings?
> Press CTRL-J/CTRL-K to move down/up and preview contents.

Reply:
In my configuration, there are key bindings set for <C-J> and <C-K>. How can I utilize these two shortcuts while using 'fzf'?
<img width="1460" alt="image" src="https://github.com/linrongbin16/fzfx.nvim/assets/23580595/057e1d8c-3c65-4922-9e6c-0b526cb3f2df">

Reply:
I think you can simply skip CTRL-J/CTRL-K keys.

Just see if the outgoing/incoming calls results are expected, without exceptions.

Reply:
Ok, It works well in my macOS.